src/main/java/picard/sam/SplitSamByNumberOfReads.java

```diff
@@ -0,0 +1,168 @@
+    static final String USAGE_DETAILS = "This tool splits the input query-grouped SAM/BAM file into multiple BAM files " +
+            "while maintaining the sort order. This can be used to split a large unmapped BAM in order to parallelize alignment." +
+            "It will traverse the bam twice unless TOTAL_READS_IN_INPUT is provided." +
+            "<br />" +
+            "<h4>Usage example:</h4>" +
+            "<pre>" +
+            "java -jar picard.jar SplitSamByNumberOfReads \\<br />" +
+            "     I=paired_unmapped_input.bam \\<br />" +
+            "     OUTPUT=out_dir \\ <br />" +
+            "     TOTAL_READS_IN_INPUT=800000000 \\ <br />" +
+            "     SPLIT_TO_N_READS=48000000" +
+            "</pre>" +
+            "<hr />";
+    @Argument(doc = "Input SAM/BAM file to split", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Argument(shortName = "N_READS", doc = "Split to have approximately N reads per output file.", mutex = {"SPLIT_TO_N_FILES"})
+    public int SPLIT_TO_N_READS;
+
+    @Argument(shortName = "N_FILES", doc = "Split to N files.", mutex = {"SPLIT_TO_N_READS"})
+    public int SPLIT_TO_N_FILES;
+
+    @Argument(shortName = "TOTAL_READS", doc = "Total number of reads in the input file. If this is not provided, the " +
+            "input will be read twice, the first time to get a count of the total reads.", optional = true)
+    public long TOTAL_READS_IN_INPUT;
+
+    @Argument(shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME, doc = "Directory in which to output the split BAM files.")
+    public File OUTPUT;
+
+    @Argument(shortName = "OUT_PREFIX", doc = "Output files will be named <OUT_PREFIX>_N.bam, where N enumerates the output file.")
+    public String OUT_PREFIX = "shard";
+
+    private final Log log = Log.getInstance(SplitSamByNumberOfReads.class);
+
+    protected int doWork() {
+        IOUtil.assertFileIsReadable(INPUT);
+        IOUtil.assertDirectoryIsWritable(OUTPUT);
+        final SamReaderFactory readerFactory = SamReaderFactory.makeDefault();
+        final SamReader reader = readerFactory.referenceSequence(REFERENCE_SEQUENCE).open(INPUT);
+        final SAMFileHeader header = reader.getFileHeader();
+
+        if (header.getSortOrder() == SAMFileHeader.SortOrder.coordinate) {
+            log.warn("Splitting a coordinate sorted bam may result in invalid bams " +
+                    "that do not always contain each read's mate in the same bam.");
+        }
+        if (!header.getVersion().equals(SAMFileHeader.CURRENT_VERSION)) {
+            log.warn(String.format("Input file's version is %s, but the current SAM format version is %s. Outputs will be written " +
+                    "with current version.", header.getVersion(), SAMFileHeader.CURRENT_VERSION));
+        }
+
+        final ProgressLogger firstPassProgress = new ProgressLogger(log, 1000000, "Counted");
+        if (TOTAL_READS_IN_INPUT == 0) {
```

Comment:
Can you check that input file is a regular file to avoid eating-up a stream?

Reply:
I don't know how to check if a file is a stream directly, but I put an error and non-zero return code before the first pass iteration begins. It should still fail quickly with a more informative error message.